### DISCUSSION TITLE
Unable to select  Charger Use Mode

### DISCUSSION THREAD

**Discussion by stringkillersa** [wills106/homeassistant-solax-modbus#39] [Q&A] [answered]
Hello, 
I have a Solax X3, G4 from 2021
Connected using TCP/IP => RS485 Modbus converter

I can read all entities, it works beautiful.
But, I cannot change any settings. I have tried these entities

charger_use_mode (Feedin Priority/Back Up Mode/Manual Mode)
selfuse_night_charge_enable (Disabled/Enabled)
manual_mode_select (Stop Charge and Discharge/Force Charge/Force Discharge)

But none of them changes anything in the inverter.
Next reading just returns previous setting.
Nothing reported in the HA logs.

Am I missing something?
Are there any settings in the inverter that can block changes?
Any ideas are welcome!

See below for one test setup.

alias: Test Selfuse/Manual
trigger:
  - platform: webhook
    webhook_id: testa-selfuse-feedin-XyDo7ocZ-Lp2-nmhSZ1vYRP4
action:
  - service: select.select_option
    data:
      option: Manual Mode
    target:
      entity_id: select.solax_1_charger_use_mode
mode: single

**Comment by infradom** [comment]
Hello, I have the same type of inverter, but using the serial modbus connection. I have ordered a TCP/Modbus converter, but this may take a couple of weeks to arrive.

I can change the use mode and other parameters , on one condition: the inverter should not be in sleep mode, so I can only change this when sunlight comes in. This is a known solax issue

**Reply by stringkillersa** [reply]
Mine is in working mode, full sunlight.

**Comment by wills106** [comment, marked as answer]
When you change the work mode on the inverter or app/cloud does select.solax_charger_use_mode change?

**Reply by stringkillersa** [reply]
Yes, I just changed mode on the front panel, and it is reflected into HA.
I have no Ethernet or WIFI adapter, thus no Cloud service either.

**Reply by wills106** [reply]
Have you tried to manually change the mode on the Lovelace dashboard. Or only through an automation?

**Reply by stringkillersa** [reply]
I have tested both. Using select on dashboard and using simple automation (see my example). The entity is changed until next communication 15s later and is restored to previous value.

**Reply by stringkillersa** [reply]
https://user-images.githubusercontent.com/100370500/159485297-87ef2567-cef5-4ab7-ae7c-f55b60f03df5.mp4

Here is a videosnippet showing the test/change on the dashboard and the value reverting.

**Reply by wills106** [reply]
I am guessing it reverts back to the original when your sensor updates again.

Do you get any errors in the log? Either HA or the adapter it's self.
I am wondering if your TCP/IP => RS485 is not passing the write register requests to the inverter.

Is there an option to manually write to a register through the adapter directly? Is there a web interface?
Can you manually write 1 to 0x1F that will change the "Charger Use Mode" to "Feedin Priority"

**Reply by stringkillersa** [reply]
No errors in the HA log. I can't see that the adapter has any log, unfortunately.
It has a simple web interface and a more complex separate setup software. 
I havent found a way to log or show the data passed, nor any way to enter commands manually.
I will loook through it again...

**Reply by stringkillersa** [reply]
I have studied the converter manual more closely, and it says it is a transparant bridge.
From the manual:
"The serial port server with Modbus gateway function does not have station address and
registers. It is a communication bridge. It will generate Modbus according to the Slave ID,
function code, register number and number of registers in the Modbus TCP command sent by
the user software to the Modbus gateway. RTU designates and outputs from the serial port.
Think of it as a protocol "translator"."

**Reply by stringkillersa** [reply]
Also, a hidden Solax entity is called "lock_state".
What is this?
In my case the value is "locked"
Suspicious?

**Reply by infradom** [reply]
Mine says "unlocked"
I cannot remember that I did something to unlock, but I might have forgotten. 
I can check if my solar system installed did something to unlock it

**Reply by wills106** [reply]
I think when you login via the SolaX Cloud it unlocks it.

There is a register called "UnlockPassword" I don't know if that's to set the password for the LCD or if you write the password to it, it unlocks the inverter?

**Reply by stringkillersa** [reply]
Hm...
The inverter has two passwords, one for user settings and one for advanced settings.
I cannot find any way to remove the password, I have the password to get access through the panel.

**Reply by stringkillersa** [reply]
Oh and btw, I came across Solax Modbus documentation för X1 & X3 version V3.21, If you need it.

**Reply by wills106** [reply]
Mine is unlocked.

I still need them two passwords to do things on the Inverter LCD, I am wondering if we need to write one of the 4digit codes to the "UnlockPassword" register and then perhaps it will unlock the control for the rest?

I have seen V3.25 but thanks anyway.

**Reply by stringkillersa** [reply]
As both @infradom and @wills106 are unlocked, it seems as the lock_state could could be involved. 
It would be a good idea to test writing the passwords into that register, but how can I do this?

**Reply by wills106** [reply]
I can't guarantee what this will do, so you are doing so at your own risk!

If you open const.py you will see the following lines around line 50

```
BUTTON_TYPES = [
    ["Battery Awaken",
        "battery_awaken",
        0x56,
        1,
    ],
]
```

Change it to the following:

```
BUTTON_TYPES = [
    ["Battery Awaken",
        "battery_awaken",
        0x00,
        2014,
    ],
]
```

If 2014 doesn't work try 6868

**Reply by stringkillersa** [reply]
@wills106 YES! We were right on track here, and you just unlocked the holy grail ;)
It was the default password 2014 and the lock_state entity turned to unlocked. 
And yes, I can change from Self_use to manual mode now.
So maybe you could add the register "UnlockPassword" as an entity in the future :)
BIG THANKS!

Another related thing: 
If I change mode to "Manual" and then select "force discharge", it works OK - 
BUT - if I then select "Stop Charge and Discharge" - it immediately begins charging instead.
I had expected it to stop and idle.
Any clues?

It might be a behavior related to some other settings...

**Reply by wills106** [reply]
Glad we got that sorted! Now we know the Inverter never needs to be connected to the cloud in order to gain full control.

You will probably need input from another Gen4 owner to understand how the different modes work as I only have a Gen3

**Reply by stringkillersa** [reply]
Yes, it's great with community efforts :)
Thanks again, I'll take a deep dive in all settings and do some trial and error first.

**Comment by infradom** [comment]
yes, the both changes are in sync (with a delay of max 15 seconds).
When the inverter is in sleep mode, it reverts to the previous state after 15 seconds

**Reply by wills106** [reply]
Bit off topic but does the Battery Awaken button do anything? Does it wake the inverter up while it's asleep?

At the moment my Inverter never goes to sleep as I have enough Solar to last 24hrs at the moment. So I can't test it myself.

**Reply by infradom** [reply]
I have not tested this yet, will do so in the coming days

**Comment by infradom** [comment]
sorry, you probably asked this question to @stringkillersa ...

**Comment by infradom** [comment]
I have read that the waveshare adapter has some kind of 'storage' mode, in which the adapter memorizes the previous state.
I believe you will have to disable this 'cachihng/storage' feature on the waveshare device

**Reply by infradom** [reply]
But if this is would be the cause of the problem, you should see the change in the solax app

**Reply by stringkillersa** [reply]
The Waveshare adapter is in "Simple Modbus to TCP/IP mode". 
I have tested all modes, but no change.

**Reply by infradom** [reply]
I expect to receive my waveshare adapter next week, then I can test on my Gen4 X3 system

**Reply by infradom** [reply]
In the mean time, what software version is your inverter running ?

**Reply by stringkillersa** [reply]
ARM: 2.03
DSP: 2.07

**Reply by infradom** [reply]
thanks, I am not at home now, I believe I cannot read my version numbers from the solax app.
I will verify my version numbers shown on the device when I come home.
The hidden version numbers in HA look different:
- firlmware version inverter master : 10
- Firmware version manager: 3
- Firmware modbus TCP Major: 0
- Firmware odbus TCP Minor:  6

**Reply by stringkillersa** [reply]
Firmware Version Inverter Master: 6
Firmware Version Manager: 7
Firmware Version Modbus TCP Major: 0
Firmware Version Modbus TCP Minor: 6
These are mine from HA

**Reply by infradom** [reply]
I am back home, and checked my version settings:
 ARM: 2.03
 DSP: 2.07
So these 2 parameters are is identical to yours. Firmware inverter master and version manager are different
I will now check if I can reporoduce the "Stop Charge and Discharge" error on my Gen4 system. In the source code, I do not immediately see an error ...

**Reply by wills106** [reply]
I think @stringkillersa has got most of it working now after using the "Unlock Inverter" button?

**Reply by stringkillersa** [reply]
There is no error when selecting "stop charge and discharge", it just does not stop. If it is charging and I select manual mode, then send "stop charge and discharge" - nothing really happens. It seems the charging power goes down a bit, but thats it.  As a remedy I change charge/discharge current instead now, this works immediately and has the advantage of not needing to change from Self_use_mode.

Btw, my inverter is set to 30A charge and discharge from factory, but in the modbus protocol it says maximum 25A (and thus this integration as well).  Just a note...

**Reply by wills106** [reply]
> Btw, my inverter is set to 30A charge and discharge from factory, but in the modbus protocol it says maximum 25A

SolaX keep changing Charge / Discharge between Firmware revisions. It's hard to know what to set it at in the Integration.

**Reply by infradom** [reply]
OK,  glad to hear that the 'stop change and discharge' issue is not a big problem. I also prefer to modify the current limits.
I started playing with the 'stop charge and discharge' option, but could not reach a conclusion yet as the behavior during daytime (sun) and night might be different.

I expect my waveshare adapter to arrive tomorrow, so then I can see if the other outstanding problem (blocking calls warning) also disappears in my setup.

**Reply by stringkillersa** [reply]
@infradom Another note regarding  "stop charge and discharge" By looking at the latest Inverter (X3/G4) manual, there is no "manual mode" no longer. Its now called "Force Time Use" and  "stop charge and discharge" does not exist.  So there are constant changes here.. and the modbus docs does not keep up either. I got the most "up to date" docs for modbus directly from Solax yesterday, but that one is version-wise older than the one circling around Internet...

**Reply by wills106** [reply]
So the Gen4 docs is older than V3.07?

**Reply by stringkillersa** [reply]
@wills106 The modbus docs I recieved from Solax yesterday is V3.19, but i found 3.21 on the internet...

**Reply by wills106** [reply]
The v3.21 and v3.25 I have seen on the internet are for the Gen3. The Gen4 docs are seperate. Is the v3.19 for the Gen3 or Gen4?

**Reply by stringkillersa** [reply]
This (v3.19)was supposed to cover latest hybrid versions. The last date of comments in the "history list" is Jan 29th of 2022. The only mention of G4 is Jan 4th, 2022 "Upgrade Example For X1G4(File DSP) Upgrade"

**Reply by wills106** [reply]
Yea sounds like that's for the Gen4 and is newer than the v3.21 and v3.25

**Reply by wills106** [reply]
Are you on community.home-assistant.io ?

**Reply by infradom** [reply]
Gen3 and Gen4 user manuals and modbus docs have different version numbering schemes. I am fairly sure that you are reading a Gen3 manual. The Gen4 manuals have a picture of the Gen4 front (single white panel).
Regarding the stop charge and discharge / manual mode:

I believe we have to interpret this as "stop charging and discharging from GRID"
So this explains why the device still charges from solar.
I also interpret "Force Charge" as "Force charge from grid" and "Force Discharge" as "Force discharge to Grid"

**Reply by stringkillersa** [reply]
@infradom I requested documentation for X1-X3 G4, and this (v3.19) was the latest they said. The V3.21 sems to relate to X3 as revision dates stop 2019, but v3.19  actually seems correctly relate to G4 as revison dates are until 2022. I recieved this document agreed on developement purpose only and not for public share, so please send your email to me at stig.andersson@ymex.se and I'll reply with the doc.
@wills106 I'm not on community.home-assistant.io

**Comment by stringkillersa** [comment]
What is your Solax  "Lock_state"?
Is it also "Unlocked"?
Mine is "Locked"
If this have anything or nothing to do with it...

**Reply by infradom** [reply]
As mentioned before, my Gen4 X3 lock state is "unlocked"

**Comment by infradom** [comment]
Ok, I wiil work on this next week (I am a gen4 owner)


Op wo 23 mrt. 2022 10:26 schreef Stig Andersson ***@***.***>:

**Comment by infradom** [comment]
I believe youy are reading a Gen3 user manual. If not, can you share the
document?

Thanks

On Mon, Mar 28, 2022 at 8:42 AM Stig Andersson ***@***.***>
wrote:

> @infradom <https://github.com/infradom> Another note regarding "stop
> charge and discharge" By looking at the latest Inverter (X3/G4) manual,
> there is no "manual mode" no longer. Its now called "Force Time Use" and
> "stop charge and discharge" does not exist. So there are constant changes
> here.. and the modbus docs does not keep up either. I got the most "up to
> date" docs for modbus directly from Solax yesterday, but that one is
> version-wise older than the one circling around Internet...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wills106/homsassistant-solax-modbus/discussions/39#discussioncomment-2449715>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2B23UEUERJKFQPW245NJ3VCFIDXANCNFSM5RKRVODQ>
> .
> You are receiving this because you were mentioned.Message ID:
> <wills106/homsassistant-solax-modbus/repo-discussions/39/comments/2449715@
> github.com>
>


-- 
Jan Colpaert
+32 478 882435

**Comment by JBayangosB** [comment]
Hey @stringkillersa my system seems to have similar options to force_discharge in manual mode but unfortunately I can't seem to get it to discharge, are you okay to share what settings you needed to get it done?
